### PR TITLE
[new opmath] Deprecation warnings and making `LinearCombination.compare` more user friendly

### PR DESCRIPTION
**Pull request by Qottmann**
Removing `LinearCombination._obs_data`, which requires updating `LinearCombination.compare`. While at it, also updating the doc string to make it clear.

**Potential Drawbacks**

A comparison with an `Observable` that has no pauli_rep (i.e. involving a `Hadamard`) will not work due to https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/functions/equal.py#L160. I consider this a separate issue for updating qml.equal accordingly. Alternative solution is to give `Hadamard` a `pauli_rep`.

```
op1, op2 = (qml.ops.LinearCombination([1.0], [qml.Hadamard(0)]), qml.Hadamard(0))
op1, op2 = (qml.ops.LinearCombination([1.0], [qml.Hadamard(0) @ qml.Hadamard(1)]), qml.Hadamard(0) @ qml.Hadamard(1))
op1, op2 = (qml.ops.LinearCombination([1.0], [qml.Hadamard(0) @ X(1)]), qml.Hadamard(0) @ X(1))
```

[sc-59339]

### CODE REVIEW

File: pennylane/ops/op_math/linear_combination.py [hunk 1]
```diff
@@ -341,6 +341,10 @@ def _obs_data(self):
         {(1, frozenset({('Prod', <Wires = [0, 1]>, ())})),
          (1, frozenset({('PauliZ', <Wires = [0]>, ())}))}
         """
+        warnings.warn(
+            "Accessing _obs_data is deprecated. You can still use LinearCombination.compare to check mathematical equivalence with other operators.",
+            qml.PennyLaneDeprecationWarning,
+        )
```
Comment (Qottmann) [inline pennylane/ops/op_math/linear_combination.py:347]: @albi3ro do you think it is still necessary to do a deprecation cycle for a void private property, `_obs_data`? Or can we just go ahead and remove it all-together?

I dont really know where this is used other than `qml.Hamiltonian.compare`, and I changed that method (in `LinearCombination`) such that it is not using this private attribute anymore.

File: pennylane/ops/op_math/linear_combination.py [hunk 2]
```diff
@@ -321,58 +321,15 @@ def simplify(self, cutoff=1.0e-12):
...
 
         .. Warning::
 
-            The compare method does **not** check if the matrix representation
-            of a :class:`~.Hermitian` observable is equal to an equivalent
-            observable expressed in terms of Pauli matrices, or as a
-            linear combination of Hermitians.
-            To do so would require the matrix form of LinearCombinations and Tensors
-            be calculated, which would drastically increase runtime.
+            While this method does not compute explicit matrices, it is still relatively expensive as it involves simplification of both operators.
```
Comment (trbromley) [inline pennylane/ops/op_math/linear_combination.py:332]: We can't try to use the Pauli representations, if available?

File: tests/ops/op_math/test_linear_combination.py [hunk 3]
```diff
@@ -705,50 +730,11 @@ def test_simplify_while_queueing(self):
...
-
-        expected = {
-            (0.5, frozenset({("Prod", qml.wires.Wires([2, 1]), ())})),
-            (1.0, frozenset({("PauliZ", qml.wires.Wires(0), ())})),
-            (1.0, frozenset({("Prod", qml.wires.Wires([0, 1]), ())})),
-        }
-
-        assert data == expected
-
-    def test_data_gell_mann(self):
```
Comment (trbromley) [inline tests/ops/op_math/test_linear_combination.py:726]: I have to admit I hadn't thought about this before - qutrit operator arithmetic is also all covered by our new approach?

**Review by trbromley — APPROVED** [review]
Thanks @Qottmann!

File: pennylane/ops/op_math/linear_combination.py (hunk 1 above)
Comment (Qottmann) [reply pennylane/ops/op_math/linear_combination.py]: After offline discussion decided to just remove this private property

File: pennylane/ops/op_math/linear_combination.py (hunk 2 above)
Comment (Qottmann) [reply pennylane/ops/op_math/linear_combination.py:332]: That is what it is doing :)!

File: tests/ops/op_math/test_linear_combination.py (hunk 3 above)
Comment (Qottmann) [reply tests/ops/op_math/test_linear_combination.py:726]: I dont think we removed any qutrit related tests so I would assume yes. This test here that I remove is just about the `_obs_data`

File: pennylane/ops/op_math/linear_combination.py [hunk 4]
```diff
@@ -404,16 +361,25 @@ def compare(self, other):
...
                 return pr1 == pr2
 
-            if isinstance(other, (LinearCombination, qml.ops.Hamiltonian)):
+            if isinstance(other, LinearCombination):
                 op1 = self.simplify()
                 op2 = other.simplify()
-                return op1._obs_data() == op2._obs_data()  # pylint: disable=protected-access
-
-            if isinstance(other, (Tensor, Observable)):
+                return qml.equal(op1, op2)
```
Comment (astralcai) [inline pennylane/ops/op_math/linear_combination.py:367]: This branch can probably be removed, since it's just the default behaviour written bellow.

**Review by astralcai — COMMENTED** [review]
Looks good! Just a small comment.

**Review by mudit2812 — APPROVED** [review]

File: pennylane/ops/op_math/linear_combination.py (hunk 4 above)
Comment (Qottmann) [reply pennylane/ops/op_math/linear_combination.py:367]: nice catch, thanks!

File: pennylane/ops/op_math/linear_combination.py (hunk 2 above)
Comment (trbromley) [reply pennylane/ops/op_math/linear_combination.py:332]: Is comparing the Pauli representation expensive? I thought it was already in a simplified representation?

Comment (Qottmann) [reply pennylane/ops/op_math/linear_combination.py:332]: Ah right now I get your point. Maybe I should say this _can_ be relatively expensive (in the cases where there is no pauli_rep) or perhaps leave out this comment about cost of this all together? Do you prefer a little detail or keeping it simple?

Comment (trbromley) [reply pennylane/ops/op_math/linear_combination.py:332]: Thanks! Maybe a sentence or two more detail can't hurt.

**Pull request merged**